chainindex/reconcile.go

```diff
@@ -0,0 +1,211 @@
+		// is the minimum non-reverted height in the index
+		for currTs != nil && currTs.Height() >= reconciliationEpoch {
+			tsKeyCidBytes, err := toTipsetKeyCidBytes(currTs)
+			if err != nil {
+				return xerrors.Errorf("failed to compute tipset cid: %w", err)
+			}
+
+			var exists bool
+			err = tx.StmtContext(ctx, si.hasNonRevertedTipsetStmt).QueryRowContext(ctx, tsKeyCidBytes).Scan(&exists)
+			if err != nil {
+				return xerrors.Errorf("failed to check if tipset exists and is not reverted: %w", err)
+			}
+
+			if exists {
+				// found it!
+				reconciliationEpoch = currTs.Height() + 1
+				log.Infof("Found matching tipset at height %d, setting reconciliation epoch to %d", currTs.Height(), reconciliationEpoch)
+				break
+			}
+
+			if len(missingTipsets) < si.maxReconcileTipsets {
+				missingTipsets = append(missingTipsets, currTs)
+			}
+			// even if len(missingTipsets) >= si.maxReconcileTipsets, we still need to continue the walk
+			// to find the final reconciliation epoch so we can mark the indexed tipsets not in the main chain as reverted
+
+			if currTs.Height() == 0 {
+				log.Infof("ReconcileWithChain reached genesis but no matching tipset found in index")
+				break
+			}
+
+			parents := currTs.Parents()
+			currTs, err = si.cs.GetTipSetFromKey(ctx, parents)
+			if err != nil {
+				return xerrors.Errorf("failed to walk chain: %w", err)
+			}
+		}
+
+		if currTs.Height() == 0 {
+			log.Warn("ReconcileWithChain reached genesis without finding matching tipset")
+		}
+
+		// mark all tipsets from the reconciliation epoch onwards in the Index as reverted as they are not in the current canonical chain
+		log.Infof("Marking tipsets as reverted from height %d", reconciliationEpoch)
+		result, err := tx.StmtContext(ctx, si.updateTipsetsToRevertedFromHeightStmt).ExecContext(ctx, int64(reconciliationEpoch))
+		if err != nil {
+			return xerrors.Errorf("failed to mark tipsets as reverted: %w", err)
+		}
+		rowsAffected, err := result.RowsAffected()
+		if err != nil {
+			return xerrors.Errorf("failed to get number of rows affected: %w", err)
+		}
+		log.Infof("Marked %d tipsets as reverted from height %d", rowsAffected, reconciliationEpoch)
+
+		// also need to mark events as reverted for the corresponding inclusion tipsets
+		if _, err = tx.StmtContext(ctx, si.updateEventsToRevertedFromHeightStmt).ExecContext(ctx, int64(reconciliationEpoch-1)); err != nil {
+			return xerrors.Errorf("failed to mark events as reverted: %w", err)
+		}
+
+		return si.applyMissingTipsets(ctx, tx, missingTipsets)
+	})
+}
+
+func (si *SqliteIndexer) backfillEmptyIndex(ctx context.Context, tx *sql.Tx, head *types.TipSet) error {
+	currTs := head
+	var missingTipsets []*types.TipSet
+
+	log.Infof("backfilling empty chain index from head height %d", head.Height())
+	var err error
+
+	for currTs != nil && len(missingTipsets) < si.maxReconcileTipsets {
+		missingTipsets = append(missingTipsets, currTs)
```

Comment:
Oh, this could get really big eh? what if they set `MaxReconcileTipsets` to fevm depth, currently `3031379`, or even max chain depth, currently `4262999`? Them's big numbers for a slice to accumulate tipsets in and then pass down into `applyMissingTipsets`.

Maybe `applyMissingTipsets` should be `applyMissingTipset` and its caller is responsible for looping?

Reply:
also, we could be making a massive list here to feed into `applyMissingTipsets` but it's only running splitstore so gets through 2 days worth and bails, so we wasted an entire chain walk and massive slice containing the tipsets

Reply:
remember that a `TipSet` also contains all the block headers, it's not small: https://github.com/filecoin-project/lotus/blob/da7ec3eedcee64fbbdcf752645c94cb6d93a29d3/chain/types/tipset.go#L21-L26

Reply:
@rvagg Changed the `backfillEmptyIndex` code to only call `applyMissingTipsets ` in batches of 1000. Given with the indexing is idempotent, the off by one of indexing the tipset which falls at boundaries of 1000 is okay.

Reply:
@aarshkshah1992 can you confirm that you refactored this again to not need any batches at all and it does it a tipset at a time?